libs/error-stack/src/report.rs

```diff
@@ -480,6 +483,87 @@ impl<C> Report<C> {
+    /// wants to handle the error, the context can then be used to directly access the context's
+    /// type. This is only possible for the latest context as the Report does not have multiple
+    /// generics as this would either require variadic generics or a workaround like tuple-list.
+    ///
+    /// This is one disadvantage of the library in comparison to plain Errors, as in these cases,
+    /// all context types are known.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// # use std::{fs, path::Path};
+    /// # use error_stack::Report;
+    /// use std::io;
+    ///
+    /// fn read_file(path: impl AsRef<Path>) -> Result<String, Report<io::Error>> {
+    ///     # const _: &str = stringify! {
+    ///     ...
+    ///     # };
+    ///     # fs::read_to_string(path.as_ref()).map_err(Report::from)
+    /// }
+    ///
+    /// let report = read_file("test.txt").unwrap_err();
+    /// let io_error = report.current_context();
+    /// assert_eq!(io_error.kind(), io::ErrorKind::NotFound);
+    /// ```
+    #[must_use]
+    pub fn current_context(&self) -> &C
+    where
+        C: Send + Sync + 'static,
+    {
+        self.downcast_ref().unwrap_or_else(|| {
+            // Panics if there isn't an attached context which matches `T`. As it's not possible to
+            // create a `Report` without a valid context and this method can only be called when `T`
+            // is a valid context, it's guaranteed that the context is available.
+            unreachable!(
+                "Report does not contain a context. This is considered a bug and should be \
+                reported to https://github.com/hashintel/hash/issues/new/choose"
+            );
+        })
+    }
+
+    /// Converts this `Report` to an [`Error`].
+    #[must_use]
+    #[cfg(any(feature = "std", rust_1_81))]
+    pub fn into_error(self) -> impl Error + Send + Sync + 'static
+    where
+        C: 'static,
+    {
+        crate::error::ReportError::new(self)
+    }
+
+    /// Returns this `Report` as an [`Error`].
+    #[must_use]
+    #[cfg(any(feature = "std", rust_1_81))]
+    pub fn as_error(&self) -> &(impl Error + Send + Sync + 'static)
+    where
+        C: 'static,
+    {
+        crate::error::ReportError::from_ref(self)
+    }
+}
+
+impl<C: ?Sized> Report<C> {
+    /// Retrieves the current frames of the `Report`, regardless of its current type state.
+    ///
+    /// You should prefer using [`Report::current_frame`] or [`Report::current_frames`] instead of
+    /// this function, as those properly interact with the type state of the `Report`.
+    ///
+    /// # Use Cases
+    ///
+    /// This function is primarily used to implement traits that require access to the frames,
+    /// such as [`Debug`]. It allows for code reuse between `Report<C>` and `Report<[C]>`
+    /// implementations without duplicating logic.
+    #[must_use]
+    pub fn current_frames_unchecked(&self) -> &[Frame] {
```

Comment:
I think `_unchecked` is misleading here. Do you have a better name? Do we want to make this private?
```suggestion
    fn current_frames_unchecked(&self) -> &[Frame] {
```
or
```suggestion
    pub(crate) fn current_frames_unchecked(&self) -> &[Frame] {
```

Reply:
I had the exact same thought as you did, but I wanted to give users the chance to implement traits regardless of if it is a single context, or multiple contexts, so I'd like to keep it public (e.g. see our implementation of `Debug`). I initially called it `leaves` or `entries` but both didn't sound quite right so any suggestion is very welcome!

Reply:
Downstream users can still implement `impl<C> Trait for Report<C>` and `impl<C> Trait for Report<[C]>`, no?

Reply:
Yea, tho it results in some duplicate code, but that should be fine I think. Then I think I'd just do `pub(crate)`. I really don't want to `pub(crate)` the frames field to be able to access it (it would be fine if we had restrictions in place, but alas, we're not here just yet)

Reply:
Yeah, that is fine. Or can you think of an (internal) facade trait which returns the same? (probably not worth it, though)